### PR TITLE
feat(cli arg): validate package names

### LINKED ISSUE #113: Verify that the resolved URL references fetches the exact package it's nested under in the lockfile
**Is your feature request related to a problem? Please describe.**

Yes. I realized that it's not very hard for an attacker to publish a package to a trusted repository, and therefore simply verifying that all packages are pulled from a trusted repository does not drastically increase the difficulty of performing an attack via lockfile modification. Instead of modifying the lockfile to resolve a dependency to a Github repo, they could simply modify it to point to a malicious package that they've published on a trusted repository.

Example of attack:
```
ms@^2.1.1:
  version "2.1.2"
  resolved "https://registry.yarnpkg.com/malicious-package/-/malicious-package-2.1.2.tgz#d09d1f357b443f493382a8eb3ccd183872ae6009"
  integrity sha512-sGkPx+VjMtmA6MX27oA4FBFELFCZZ4S4XqeGOXCv68tT+jb3vk/RyaKWP0PTKyWtmLSM0b+adUTEvbs1PEaH2w==
```

**Describe the solution you'd like**

An additional check that would increase the difficulty significantly of an attack would be to check that the package being pulled from the trusted repository has the same name as that specified in the lockfile. That way, the attacker would be forced to compromise the legitimate package itself.

The solution I'm proposing is to add an additional verifier that checks that the package name in the lockfile key matches the package name in the resolved URL.

**Describe alternatives you've considered**
N/A

### PR DESCRIPTION
**Pull request by fabioberger**
## Description

Instead of modifying the lockfile to resolve a dependency to a Github repo, they could simply modify it to point to a malicious package that they've published on a trusted repository.

Example of attack:
```
ms@^2.1.1:
  version "2.1.2"
  resolved "https://registry.yarnpkg.com/malicious-package/-/malicious-package-2.1.2.tgz#d09d1f357b443f493382a8eb3ccd183872ae6009"
  integrity sha512-sGkPx+VjMtmA6MX27oA4FBFELFCZZ4S4XqeGOXCv68tT+jb3vk/RyaKWP0PTKyWtmLSM0b+adUTEvbs1PEaH2w==
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Related Issue

Fixes #113 

Sorry I didn't wait for discussion before implementing this 😅 

## Motivation and Context

It makes it much harder for an attacker to point a legitimate dependency to a repository they control (even if on an official repository).

## How Has This Been Tested?

I've extended the existing tests to cover the new functionality. 

## Checklist:

- [ ] I have updated the documentation (if required).
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I added a picture of a cute animal cause it's fun

### CODE REVIEW

**Comment by lirantal** [comment]
Hi @fabioberger, thanks for working on this enhancement and adding the pull request with tests. I highly appreciate that ❤️ 

A few things I'd like to raise before we move forward with this suggested enhancement:

1. Are the package names consistent with the URL?

```
ms@^2.1.1:
  version "2.1.2"
  resolved "https://registry.yarnpkg.com/malicious-package/-/malicious-package-2.1.2.tgz#d09d1f357b443f493382a8eb3ccd183872ae6009"
```

Meaning, can you expect that all packages on the registry conform to their package name string such as `ms` in this case, will have the exact same URL parameter name of `ms`? Could you run some tests on bigger package lock files to ensure it?

2. We should also take into account package names using scoped namespaces. Can you update the code here to support these, and also add relevant test cases for them?

File: packages/lockfile-lint-api/__tests__/validators.packageNames.test.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        }
+      ]
+    }
+    expect(validator.validate()).toEqual(expectedValidationResult)
+  })
+
+  it('validator should succeed if all resource URLs are for correct packages', () => {
+    const mockedPackages = {
+      '@babel/code-frame@^0.16.0': {
+        resolved: 'https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.0.0.tgz'
```
Comment (fabioberger) [inline packages/lockfile-lint-api/__tests__/validators.packageNames.test.js:47]: Here we test that namespaced packages (e.g., @babel/*) work as expected.

File: packages/lockfile-lint-api/src/validators/ValidatePackageNames.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      try {
+        const packageResolvedURL = new URL(packageMetadata.resolved)
+        const path = packageResolvedURL.pathname
+        const packageNameFromResolved = path.split('/-/')[0].slice(1)
+
+        // Remove versioning info from packageName. The @ sign is the delimiter, but could also be the
+        // first character of a scoped package name. We handle this edge-case here.
+        const nameOnly = packageName.startsWith('@')
+          ? `@${packageName.slice(1).split('@')[0]}`
+          : packageName.split('@')[0]
```
Comment (fabioberger) [inline packages/lockfile-lint-api/src/validators/ValidatePackageNames.js:34]: Here is the logic to support scoped / namespaced packages.

**Comment by fabioberger** [comment]
Thanks for the quick response @lirantal 🙏 

Re: supporting namespaced packages, this is already implemented. I've commented on the relevant lines above. I've also gone ahead and tested this feature on three separate Yarn.lock files with ~18000 lines and did not hit any further edge-cases.

As for whether all `resolved` URLs will have the package name as the first part of the URL path, this is the case for the official Yarn and NPM package registries. It does not hold for arbitrary `resolved` URLs since it can contain any valid URL (e.g., `https://codeload.github.com/webpack/tooling/tar.gz/2e849403e5608b110e869b599442f89d7004e920` from Webpack's yarn.lock). It might make sense to enforce that the user specify `--allowed-hosts` to be either `npm` or `yarn` in order to use this feature. Happy to add this if you feel it's necessary.

Relevant documentation for Yarn and npmjs:
- https://classic.yarnpkg.com/lang/en/docs/yarn-lock/
- https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json#resolved

**Comment by fabioberger** [comment]
Friendly bump on this @lirantal 🙏

**Comment by lirantal** [comment]
Noted. I haven't read yet through the replies but will do my best to get to it tomorrow. Appreciate the patience here.

File: packages/lockfile-lint-api/__tests__/validators.packageNames.test.js (hunk 1 above)
Comment (lirantal) [reply packages/lockfile-lint-api/__tests__/validators.packageNames.test.js:47]: Ahh yes. While that's not the mocked package, I think it goes through the same code path to extract the @.

**Comment by lirantal** [comment]
Thank you for all of the work spent to implement this @fabioberger, wonderful job!

>  It might make sense to enforce that the user specify --allowed-hosts to be either npm or yarn in order to use this feature. Happy to add this if you feel it's necessary.

I'd appreciate if we could indeed add this extra check when someone uses `--validate-package-names` to make sure users aren't left confused why does it fail in case they have some packages resolved out of GitHub gists, artifactory remote URLs or something else that's a bit out of the regular conventions.

**Review by lirantal — CHANGES_REQUESTED** [review]

File: packages/lockfile-lint/src/config.js [hunk 3]
```diff
@@ -45,6 +45,13 @@ module.exports = (argv, exitProcess = false, searchFrom = process.cwd()) => {
         type: 'boolean',
         describe: 'validates the use of HTTPS as protocol schema for all resources'
       },
+      n: {
+        alias: ['validate-package-names'],
+        type: 'boolean',
+        describe:
+          "validates that the resource URL specifies the same package name as that listed as the lockfile entry's key",
+        implies: 'allowed-hosts'
```
Comment (fabioberger) [inline packages/lockfile-lint/src/config.js:53]: @lirantal Yargs now enforces that if package names are validated, the developer has also specified specifically allowed hosts. All hosts I know of follow the checked convention (i.e., Yarn, NPM, Verdaccio). Some might not, but that is out-of-scope for this PR. At least now, users who _don't_ specify a host and use non-standard sources (e.g., Github, etc.) will get a friendly message notifying them that package name validation only works when particular hosts have been chosen.

**Comment by lirantal** [comment]
@fabioberger I'm commuting now home so will merge it later tonight.

**Comment by lirantal** [comment]
looks like npmjs is down, GitHub Actions CI fails to publish: 
image

I'll give it a spin tomorrow.

**Comment by fabioberger** [comment]
Thanks for merging @lirantal! Please let me know once you've been able to publish 🙏

**Comment by lirantal** [comment]
New versions are out :-)

**Pull request merged**